build/bazel/remote/execution/v2/remote_execution.proto

```diff
@@ -1724,6 +1760,11 @@ message CacheCapabilities {
   // Note that this does not imply which if any compressors are supported by
   // the server at the gRPC level.
   repeated Compressor.Value supported_compressor = 6;
+
+  // Whether compression of inlined data is also supported (if so, the
+  // server must support each of the compressors specified in the
+  // `supported_compressor` field).
+  bool inlined_compressed_blobs = 7;
```

Comment:
Relatedly, I also think that this phrasing far too strict. With the repeated field that you're adding, you've already introduced a two-way method for negotiating compression algorithms on a per-request basis.

Do we even need this flag? Why not always allow clients to just provide a list of compression functions, regardless of whether the server does anything with it?

Reply:
It is possible to upload blobs with bytestream in parallel, but it's often simpler to do it in a single RPC call, and batching small requests together can reduce certain overheads (it's a balancing act as always).

Reply:
Have you got numbers on what the overhead is?

Reply:
I don't have any useful numbers from my own setup to share (I'm currently using gRPC level compression for these unary RPCS with inlined data, though I would prefer to migrate to REAPI-level compression).

Reply:
For most use-cases it doesn't matter, but when uploading lots of small blobs it can be very material. With e.g. 1kb blobs, you're likely to be parallelism bound (100 RPCs in flight / connection -> 1000 blobs/second/connection or so); batch uploads are then very necessary if you've got thousands of blobs to upload.

For the field comment: should this apply to _all_ inline data, or just be scoped to client-sent data? I.e. I'm wondering about changing it to "Whether compression of client-sent inlined data is also supported". That'd allow a server to support returning compressed data (via client-specified acceptable_compressors) without having to support accepting it.

Though...maybe that's a bad thing, and we should keep the existing semantics so as to introduce less variation in servers.

Reply:
Updated the comment. This only applies to client-sent data since we have allowed the server to always return uncompressed data if it chooses to.  It would be toothless to state here that servers must also support this for downloads.

Jumping back to @EdSchouten's question...
> What's the use of uploading inlined blobs?

This (small) PR doesn't add support for inlining blobs in new places. It just allows compressed inlined data where we previously only allowed uncompressed inlined data. Which IMO is a net benefit for servers that choose to support inlined data for uploads.

Reply:
> This (small) PR doesn't add support for inlining blobs in new places. It just allows compressed inlined data where we previously only allowed uncompressed inlined data.

I get that. But there's also protocol evolution, right? Does it make sense to extend features that would in retrospect be considered redundant/unnecessary? If we're adding support for compressed uploads in BatchUploadBlobs(), does it even make sense to invest in extending UpdateActionResult() to support it?

Reply:
IMO this is a useful improvement for UpdateActionResult rather than something we would want to discourage. On the server side it is probably not much harder to implement this for both messages than for only BatchUpdateBlobs.

I wouldn't be opposed to advertising the capabilites of each rpc call separately, but I'm not sure if that would help any server imlementations in practice.

Reply:
Upon re-reading the API, I'm actually leaning towards Ed's suggestion of prohibiting compressed blob inlining in UpdateActionResult. While it's hopefully not *that* hard, it is a new set of fields and whatnot to support compression on (clients don't *set* stdout/stderr fields anywhere else, output contents, etc; Servers don't usually read it), and allowing it would force servers to support rewriting compressed fields on the way out (e.g. if one client uploads an action result with inlined compressed data, and a second client calls GetActionResult and does not specify compression being allowed.)

Of course, if anyone actually cares to have support of compressed+inlined uploads, that's a different story entirely. But right now it sounds like it's unmotivated by concrete use-cases, and mostly desirable for symmetry? Which I care about as well, though perhaps the tax in this case outweighs it.

Reply:
Don't servers already need to support modifying update ActionResult messages though? There's nothing preventing clients and workers from uploading ActionResult messages with inlined uncompressed data, and servers must not inline this data unless requested by the client in response to GetActionResult messages.

The strategy I implemented in bazel-remote was to accept ActionResult messages with possibly inlined data, store any inlined blobs separately in the CAS since they may be requested separately, de-inline them from the ActionResult before storing that in the ActionCache, and then returning that possibly modified Actionresult.

This allows clients and workers to upload the results of executing an action in a single RPC call in the best case. If we limit compressed inlined blob support to BatchUpdateBlobs for uploads, then the client/worker needs to make at least two RPC calls if you want to use compression (but not gRPC level compression). Even if they're not sequential it's simpler to do this in a single RPC.